packages/core/src/utils.js

```diff
@@ -767,6 +767,31 @@ export function retrieveSchema(schema, rootSchema = {}, formData = {}) {
     return resolveCondition(schema, rootSchema, formData);
   }
 
+  if (resolvedSchema.properties) {
+    const properties = {};
+
+    Object.entries(resolvedSchema.properties).forEach(entries => {
+      const propName = entries[0];
+      const propSchema = entries[1];
+      const rawPropData = formData && formData[propName];
+      const propData = isObject(rawPropData) ? rawPropData : {};
+      const resolvedPropSchema = retrieveSchema(
+        propSchema,
+        rootSchema,
+        propData
+      );
+
+      properties[propName] = resolvedPropSchema;
+
+      if (
+        propSchema !== resolvedPropSchema &&
+        resolvedSchema.properties !== properties
```

Comment:
It looks like `properties` will always be a different object than `resolvedSchema.properties`, since you declared a new `properties` object on line 771. So I think this would always be true.

Reply:
No, this is possible ONLY when the recursive call of retrieveSchema returns a modified schema for the specific property. Otherwise resolvedSchema will be returned unchanged.

Reply:
The idea of this check is to reduce creation of shallow copy for the retrieved schemas. So if a schema isn't computed schema (doesn't have any if/then/else) it will be returned unchanged (and reference equality check will be enough for this condition)

Reply:
Sorry if I'm being unclear, but essentially I don't think you will ever have referential equality between `resolvedSchema.properties` and `properties` because `properties` is always a new object.

Reply:
No, let me explain. Before loop we initialize `properties` variable as a new object, you right. And during the iteration (first time when we meet `propSchema !== resolvedPropSchema`) the condition `resolvedSchema.properties !== properties` will be truthy, you right too. But after we make a shallow copy of the schema with new properties container `resolvedSchema = { ...resolvedSchema, properties };` during any next iteration `resolvedSchema.properties !== properties` will be falsy and we will not make shallow copy. The bonuse is: when resolvedSchema doesn't contain properties with computed schema we do not make shallow copy at all and return the unmodified object.

Reply:
🤦‍♂️ I finally see it now! Thank you for this explanation (and for bearing with me)!